src/infrastructure/channels/reliable_channel.rs

```diff
@@ -0,0 +1,146 @@
+
+/// This channel should be used for reliable processing of packets.
+///
+/// *Details*
+///
+///  1. Reliable.
+///  2. Guarantee of delivery.
+///  3. Option to specify if data needs to be ordered.
+///  4. Packets will not be dropped.
+///  5. Duplication not possible
+pub struct ReliableChannel {
+    // settings
+    ordered: bool,
+    config: NetworkConfig,
+
+    // reliability control
+    seq_num: u16,
+    waiting_packets: LocalAckRecord,
+    their_acks: ExternalAcks,
+    dropped_packets: Vec<Box<[u8]>>,
+
+    // congestion control
+    network_quality_measurer: NetworkQualityMeasurer,
+    congestion_data: SequenceBuffer<CongestionData>,
+    quality: NetworkQuality,
+    rtt: f32,
+}
+
+impl ReliableChannel {
+    /// Create a new instance of the unreliable channel by specifying if channel needs to order incoming packets.
+    pub fn new(ordered: bool, config: NetworkConfig, ) -> ReliableChannel {
+        ReliableChannel {
+            // settings
+            ordered,
+            config: config.clone(),
+
+            // reliability control
+            seq_num: 0,
+            waiting_packets: Default::default(),
+            their_acks: Default::default(),
+            dropped_packets: Vec::new(),
+
+            // congestion control
+            network_quality_measurer: NetworkQualityMeasurer::new(config),
+            congestion_data: SequenceBuffer::with_capacity(<u16>::max_value() as usize),
+            quality: NetworkQuality::Good,
+            rtt: 0.0,
+        }
+    }
+}
+
+impl Channel for ReliableChannel {
+    /// This will prepossess an reliable packet.
+    ///
+    /// 1. Add congestion data entry so that it could be monitored.
+    /// 2. Queueing new packet for acknowledgement system.
+    /// 4. Fragmentation of the payload.
+    fn pre_process(&mut self, payload: &[u8], delivery_method: DeliveryMethod) -> NetworkResult<PacketData> {
+        if payload.len() > self.config.max_packet_size {
+            error!(
+                "Packet too large: Attempting to send {}, max={}",
+                payload.len(),
+                self.config.max_packet_size
+            );
+            Err(PacketErrorKind::ExceededMaxPacketSize)?;
+        }
+
+        // queue congestion data.
+        self.congestion_data.insert(
+            CongestionData::new(self.seq_num, Instant::now()),
+            self.seq_num,
+        );
+
+        // queue packet for awaiting acknowledgement.
+        self.waiting_packets.enqueue(self.seq_num, &payload);
+
+        // calculate size for or packet data.
+        let payload_length = payload.len() as u16; /* safe cast because max packet size is u16 */
+        let packet_data_size: u16 = Fragmentation::total_fragments_needed(payload_length, self.config.fragment_size);
+        let mut packet_data = PacketData::with_capacity(packet_data_size as usize);
+
+        let packet_type = if packet_data_size > 1 {
+            PacketTypeId::Fragment
+        }else {
```

Comment:
I hate to sound like @LucioFranco but could we have a space between the `}` and `else` ? =)

Reply:
lol, yeah sure. Once this get's approved I'll run cargo fmt before merging.